Enforce Dataframe Backend Checks

ryantwolf opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
Refactors all modules to inherit from `nemo_curator.modules.base.Module`. With this change, the base class now checks to see if the dataframe backend for the `Module` matches what is passed in, and will throw an error prior to computation if there is a mismatch. We also prove a new module `ToBackend()` that allows easy chaining of CPU and GPU modules.


## Usage
<!-- Potentially add a usage example below -->
```python
from nemo_curator import Module, Sequential, ToBackend
from nemo_curator.datasets import DocumentDataset

class CPUModule(Module):
    def __init__(self):
        super().__init__(input_backend="pandas")

    def call(self, dataset: DocumentDataset):
        dataset.df["cpu_lengths"] = dataset.df["text"].str.len()
        return dataset


class GPUModule(Module):
    def __init__(self):
        super().__init__(input_backend="cudf")

    def call(self, dataset: DocumentDataset):
        dataset.df["gpu_lengths"] = dataset.df["text"].str.len()
        return dataset

def main():
  client = get_client(cluster_type="gpu")
  # Load a dataset with a pandas CPU backend
  dataset = DocumentDataset.from_pandas(...)
  pipeline = Sequential([
    CPUModule(),
    ToBackend("cudf"),
    GPUModule(),
  ])

  curated_dataset = pipeline(dataset)
  curated_dataset.to_json(...)

if __name__ == "__main__":
  main()
```
## Checklist
<!--
Note: All commits need to be signed and signed off. This can be done via `-sS` flags while commiting
`git commit -sS -m "...."
-->
- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

sarahyurick commented on tests/test_backends.py:
```diff
@@ -0,0 +1,214 @@
+LocalCUDACluster = gpu_only_import_from("dask_cuda", "LocalCUDACluster")
+
+
+class CPUModule(Module):
```
Was wondering if we can also have some type of test with at least 1 CPU and GPU module? Such as fuzzy duplicates?

ryantwolf replied:
you mean with a real module, not just these dummy modules?

sarahyurick replied:
Yes.

sarahyurick commented on docs/user-guide/cpuvsgpu.rst:
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
+      client = get_client(cluster_type="gpu")
+
+      dataset = DocumentDataset.read_json("books.jsonl")
+      curation_pipeline = Sequential([
+          ScoreFilter(RepeatingTopNGramsFilter(n=5)),
+          ToBackend("cudf"),
+          DomainClassifier(),
+      ])
```
This is very nice!

sarahyurick commented on nemo_curator/modules/__init__.py:
```diff
@@ -22,12 +22,14 @@
+from .to_backend import ToBackend
 
 # GPU packages
 MinHash = gpu_only_import_from("nemo_curator.modules.fuzzy_dedup.minhash", "MinHash")
```
Do you think there is any way to take advantage of the module backends so that we don't need to use `gpu_only_import_from`?

ryantwolf replied:
Not quite sure what you mean. Are you asking if we could somehow determine that `MinHash` is a GPU module from its backend and then refactor this to do something different than `gpu_only_import_from`? I'm wondering what you think should be improved with these imports.

sarahyurick replied:
I guess I was wondering if it would be possible to use the backend checks in this PR to help all modules be importable by users. For example if a user with CPU only tries to do an `from nemo_curator import FuzzyDuplicates` then maybe this file should raise a more readable message which specifies that the module can only be run with the GPU?

ryantwolf replied:
I don't think it can. In `FuzzyDuplicates` for example, the import would fail before the `FuzzyDuplicates` class would be called due to the missing `dask_cudf` import. And we'd need the `FuzzyDuplicates` class to be there in order to check its backend.

praateekmahajan commented on docs/user-guide/cpuvsgpu.rst:
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
+Let's highlight some of the important parts of this example.
+
+* ``client = get_client(cluster_type="gpu")``: Creates a local Dask cluster with access to the GPUs. In order to use/swap to a cuDF dataframe backend, you need to make sure you are running on a GPU Dask cluster.
+* ``dataset = DocumentDataset.read_json("books.jsonl")``: Reads in the dataset to a pandas (CPU) backend.
```
I think we should I explicitly mention the implicit default argument of `backend="pandas"`

praateekmahajan commented on nemo_curator/filters/doc_filter.py:
```diff
@@ -81,6 +81,15 @@ def keep_document(self, scores: Any) -> bool:
             "keep_document method must be implemented by subclasses"
         )
 
+    def get_backend(self) -> str:
```
```suggestion
    def get_backend(self) -> Literal["pandas", "cudf", "any"]:
```

(what's any though?)

praateekmahajan replied:
Seems like this could be an [abstractproperty](https://docs.python.org/3/library/abc.html#abc.abstractproperty) (now abstractmethod + property). 

And instead of the abstract class defining it, IMO the inherited class should explicitly define that property to be `pandas` or whatever

ryantwolf replied:
"any" means that the module can operate on either "pands" or "cudf" as a backend. Will update to be a property instead of a method.

> And instead of the abstract class defining it, IMO the inherited class should explicitly define that property to be pandas or whatever

The reason that I have the base class start off with a default of `pandas` is that right now I believe every DocumentFilter has a backend of pandas. In my mind, it's more convenient to set the default and have a few classes override it in the future if they want. What do you think?

praateekmahajan commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
```
Can we add a test case for this? And also type hints for ddf which would be `dd.DataFrame`

ryantwolf replied:
What part of this do you want to test? I'm not quite seeing which part of this you want me to isolate.

praateekmahajan replied:
As someone reading this, it's not obviously clear what will be the output of `.__module__.split(".")[0]` of the type of  object's `_meta` attribute. So I would've proposed that you move this to a function and have tests for that and use that function here. 
However I see you've changed this to is_cudf_type(..) which is much more readable, and now we can unittest that directly that.

praateekmahajan commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
+        raise NotImplementedError("call method must be implemented by subclasses")
+
+    def _check_backend(self, ddf):
```
nit : name could be `_assert_correct_backend`

ryantwolf replied:
what do you think about `_validate_correct_backend` instead? Since there's no direct `assert` statement or error thrown I think something slightly more generic is better.

praateekmahajan commented on nemo_curator/modifiers/doc_modifier.py:
```diff
@@ -26,3 +26,12 @@ def __init__(self):
     def modify_document(self, text):
         pass
+
+    def get_backend(self) -> str:
+        """
+        The dataframe backend the filter operates on.
+        Can be 'pandas', 'cudf', or 'any'. Defaults to 'pandas'.
+        Returns:
+            str: A string representing the dataframe backend the filter needs as input
+        """
+        return "pandas"
```
Same as above on type hint + abstractproperty + require explicit definition in child class rather than abstract class

praateekmahajan commented on nemo_curator/modules/to_backend.py:
```diff
@@ -0,0 +1,25 @@
+        self.backend = backend
+
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
+        return DocumentDataset(dataset.df.to_backend(self.backend))
```
can add an explicit no-op if backend is the same if it's easy to check

ryantwolf replied:
From what I can see dask already has a no-op in its implementation for pandas, I assume the cudf one is similar. I'll keep it like this for now.

praateekmahajan commented on nemo_curator/modules/to_backend.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class ToBackend(Module):
+    def __init__(self, backend: str) -> None:
+        super().__init__(input_backend="any")
+        self.backend = backend
```
nit : type hint

praateekmahajan commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+        self.input_backend = input_backend
+
+    @abstractmethod
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
```
super nit : `call` could be slightly confusing for python readers, as it does overlap with the `__call__` dunder. What about we call it `fn` / `func` / `impl` / `implementation`?

sarahyurick replied:
Hmm maybe something like `run`, `execute`, `dispatch`? Something that still makes it clear that we are calling the core functionality of that class.

ayushdg replied:
No strong opinions: open to using `run/execute/call`.

VibhuJawa replied:
My opinion is actually to stick with `call`.  I think CrossFit call convention comes from Keras/TensorFlow (lol) Anyways, my mental model is call actually ends up calling `__call__ ` function so it lines up in my head. 

Other options can be `transform`, `process`  too.  

I slightly dont like `run`, `execute` because it has annotations of synchronous behavior in my head (but can be convinced of it too).

ryantwolf replied:
Ok given that it's from keras/tensorflow and crossfit I'm going to keep it that way for now.

praateekmahajan commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+class Module(ABC):
+    SUPPORTED_BACKENDS = ["pandas", "cudf", "any"]
+
+    def __init__(self, input_backend: str, name=None) -> None:
```
nit : type hints

praateekmahajan commented on tests/test_backends.py:
```diff
@@ -0,0 +1,222 @@
+            ]
+        )
+        result = pipeline(dataset)
+        result_df = result.df.compute()
```
add a small assert here for list(df.columns) which should be `id, text, gpu_lengths, cpu_lengths`

praateekmahajan reviewed (changes requested):
This looks great. Left a review, with mostly nits around type hints and function renaming  and some class abstraction opinions.

praateekmahajan commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+from nemo_curator.datasets import DocumentDataset
+
+
+class Module(ABC):
```
nit to rename this to `BaseModule`

ryantwolf commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,50 @@
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
+            )
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
```
I'll leave the return type unannotated for now.

ryantwolf commented:
@sarahyurick @praateekmahajan thanks for your reviews. I addressed the feedback that I immediately agreed with/understood. I left some follow up comments and questions on the other points.

ayushdg commented on docs/user-guide/cpuvsgpu.rst:
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
+      curation_pipeline = Sequential([
+          ScoreFilter(RepeatingTopNGramsFilter(n=5)),
+          ToBackend("cudf"),
+          DomainClassifier(),
```
It might be worth showing an example going the other way from cuDF to pandas via ToBackend in this short example as well.

ayushdg commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,82 @@
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
```
Would the `is_cudf_type` utility help in this case? https://github.com/NVIDIA/NeMo-Curator/blob/c3fb61dfc37d25a916384c8484a8da2252e9e105/nemo_curator/utils/gpu_utils.py#L19-L28

ayushdg commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,82 @@
+            raise ValueError(
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
+            )
+
```
In addition to `_validate_correct_backend`, it also makes sense to have something like `_validate_correct_cluster_type` that can also verify that for cuDF backends folks are using a GPU cluster. Perhaps the `getClient` api can set the cluster type in some variable? 

I realize that there is no clean way to do this today.

ryantwolf replied:
I think it does make sense to do this. We could have a global variable `CLUSTER_TYPE` that gets set during the `get_client` call. Then, in the `BaseModule` we can just check this variable. The only problem that I see with this right now though is how to handle the case where we connect to an existing Dask cluster. We'd need to make sure that it also set some environment variable or something designating the cluster type. I think we can enforce this in our example slurm scripts, but do you mind if I do this in another PR?

ayushdg replied:
Makes sense. Let's handle this in a followup. For cases where this variable is not set like when users are setting things up themselves without using our recommended scripts, we can keep the check lenient enough to still allow processing (maybe with a warning if preferred).

ayushdg commented on nemo_curator/modules/base.py:
```diff
@@ -0,0 +1,82 @@
+        backend = type(ddf._meta).__module__.split(".")[0]
+        if backend != self.input_backend:
+            raise ValueError(
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
```
Might be worth giving a hint here by mentioning that Users can use the `ToBackend` module to switch backends.

ayushdg commented on nemo_curator/modules/to_backend.py:
```diff
@@ -0,0 +1,36 @@
+            backend (str): The backend to transfer the dataset to. Can be "pandas" or "cudf"
+        """
+        super().__init__(input_backend="any")
+        self.backend = backend
```
Might be worth adding the `supported_backends` list and check here as well to ensure users don't try to convert to an unsupported backend.

ayushdg commented on tests/test_backends.py:
```diff
@@ -0,0 +1,437 @@
+        return dataset
+
+
+class AnyModule(BaseModule):
+    def __init__(self):
+        super().__init__(input_backend="any")
+
+    def call(self, dataset: DocumentDataset):
+        dataset.df["any_lengths"] = dataset.df["text"].str.len()
+        return dataset
```
repeated class

ayushdg commented on tests/test_backends.py:
```diff
@@ -0,0 +1,437 @@
+
+@pytest.fixture
+def cpu_data():
+    df = pd.DataFrame(
```
nit: You could create a common fixture that returns the original data and have two fixtures that use the common one as input for generating cpu and gpu data instead of repeating the data twice.

eg: 
```python
@pytest.fixture
def raw_data():
...

@pytest.fixture
def cpu_data(raw_data):
    return ...

@pytest.fixture
def gpu_data(raw_data):
 ...
```

ryantwolf replied:
sweet, thanks for letting me know!

ayushdg commented on tests/test_backends.py:
```diff
@@ -0,0 +1,437 @@
+        assert_eq(result_df["cpu_lengths"], gt_cpu_lengths)
+        assert_eq(result_df["gpu_lengths"], gt_gpu_lengths)
+
+    def test_wrong_backend(self, cpu_data):
```
nit: could you add the other way around as well?

ayushdg commented on tests/test_backends.py:
```diff
@@ -0,0 +1,437 @@
+            _ = result.df.compute()
+
+
+@pytest.fixture
```
same as above

sarahyurick reviewed (approved):
LGTM!

praateekmahajan reviewed (approved):
LGTM thanks for adding this!

ayushdg reviewed (approved):
lgtm! Thanks
